### PR TITLE
Intermediate form schema table to collapse dupe schemas

### LINKED ISSUE #662: Creating form draft results in 504 when there are many form fields
@lognaturel was working on a server with a large number of form versions and hundreds of thousands of rows in the `form_fields` table. Quoting @lognaturel from Slack:

> Creating a form draft takes a very long time and sometimes 504s. This is for any form on the server. Publishing a form draft generally 504s.

The strong suspicion is that the `check_field_collisions` trigger is responsible:

> I can get things snappier by deleting all the form_field entries that correspond to the form with many versions except for the one for the active form def.

> The behavior I’m experiencing is exactly what @matthew-white described at #381. So I think the now-reverted change to the trigger made it so slow the problem would show up with fewer versions.

Two ideas for improvement:

> I believe that there are improvements to make to that trigger. There actually is a `formId` column on that table so it seems like at the very least the slowness could be constrained to only forms with many versions rather than affecting everything on the server.
> (and compounding if there are many forms with lots of versions)

> I wonder whether we could detect when the form schema hasn’t changed at all.
> Seems there are lots of cases in which we’re saving a LOT of duplicate data.

### PR DESCRIPTION
**Pull request by ktuite**
Introduces an intermediate form schema table for collapsing form fields from different versions of the same form that share the same schema. Closes issue #662.

This PR is meant to be all under the hood, but it touches a lot of things under the hood. The main structure of these changes are as follows:

1. Migration to collapse duplicate form schemas. This involves adding a `form_schemas` table that just stores an int, and replacing `formDefId` columns on `form_fields` and `form_defs` with a link to this schema ID `schemaId` that can be shared across defs. There is also a block of code that goes through all existing form defs and compares their fields to assign appropriate schema IDs. 
2. A change of all queries that include `form_fields` to join with `form_defs` through the schema ID instead of the def ID. Also changes to queries for dataset property fields, which are coupled with regular form fields.
3. Changes to the logic of `forms.createNew` and `forms.createVersion` to make sure this new schema ID gets checked, created and used. The structure of the code for these two functions has been made into async/await.
4. Moving the trigger (that was the cause of the 504 in the linked issue) out of SQL and into the code. This point is still up for discussion... the trigger is meant to check for disallowed field type downcasts.

#### What has been done to verify that this works as intended?

Tests, but I'm also going to do an external round of testing. 

#### Why is this the best possible solution? Were any other approaches considered?

A lot of discussion went into possible solutions for the root cause issue #662, from getting people to start using entities, to allowing them to update their attachments on published forms, to this chosen solution of collapsing redundant form field schemas by introducing an intermediate schema representation.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

It should make form def uploads faster for users, especially in the case where they are rapidly uploading new versions with the exact same schema in order to automatically update a CSV attachment.

There regression risks (that we're trying hard to avoid) are:
- data loss from migration
- a bug in how the new form schema table is joined in various queries that breaks functionality (but this is what our integration tests are for)
- some other part of the process gets unintentionally slower

#### Does this change require updates to the API documentation? If so, please update docs/api.md as part of this PR.

No.

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes

### CODE REVIEW

File: lib/model/query/forms.js [hunk 1]
```diff
@@ -466,15 +486,26 @@ const getAllByAuth = (auth, options = QueryOptions.none) => ({ all }) =>
...
+  order by form_fields."order" asc`)
     .then(map(construct(Form.Field)));
 
 const getStructuralFields = (formDefId) => ({ all }) =>
-  all(sql`select * from form_fields where "formDefId"=${formDefId} and (type='repeat' or type='structure') order by "order" asc`)
+  all(sql`select form_fields.* from form_fields
+  join form_schemas on form_schemas."id" = form_fields."schemaId"
+  join form_defs on form_schemas."id" = form_defs."schemaId"
+  where form_defs."id"=${formDefId} and (form_fields.type='repeat' or form_fields.type='structure')
+  order by form_fields."order" asc`)
```
Comment (ktuite) [inline lib/model/query/forms.js:514]: update casing for these queries

**Review by ktuite — COMMENTED** [review]
Interactive review with @sadiqkhoja 

Another note for when applying the merge: Add a message in the squash commit message that explains the issue: too many copies of duplicate form fields when forms with the same schema are re-uploaded

File: lib/model/query/forms.js [hunk 2]
```diff
@@ -59,35 +59,34 @@ select id from form`))
...
-                  publish
-                );
-              }
-            })
-            .then(always(savedForm));
-        })));
+const createNew = (partial, project, publish = false) => async ({ run, Datasets, FormAttachments, Forms, Keys }) => {
+  const keyId = await partial.aux.key.map(Keys.ensure).orElse(resolve(null));
+  const fields = await getFormFields(partial.xml);
+  const dataset = await (partial.aux.key.isDefined() ? resolve(Option.none()) : getDataset(partial.xml)); // Don't parse dataset schema if Form has encryption key
```
[the remark this answers is not in the document]
Comment (ktuite) [reply lib/model/query/forms.js]: resolved in PR #728

File: lib/model/query/forms.js [hunk 3]
```diff
@@ -59,35 +59,34 @@ select id from form`))
...
+
+  await Forms.checkDeletedForms(partial.xmlFormId, project.id)
+    .then(Forms.rejectIfWarnings);
+
+  const savedForm = await Forms._createNew(partial, partial.def.with({ keyId }), project, publish);
+
+  const ids = { formId: savedForm.id, formDefId: savedForm.def.id };
+  await run(insertMany(fields.map((field) => new Form.Field(Object.assign(field, ids)))));
+
+  await FormAttachments.createNew(partial.xml, savedForm, partial.xls.itemsets);
```
[the remark this answers is not in the document]
Comment (ktuite) [reply lib/model/query/forms.js]: done in PR #728

File: lib/model/query/forms.js [hunk 4]
```diff
@@ -110,49 +109,46 @@ const _getDraftToken = (form) => {
...
-                return Datasets.createOrMerge(
-                  new Dataset({ name: dataset.get(), projectId: form.projectId }, { formDefId: savedDef.id }),
-                  fields.filter((field) => (field.propertyName)).map((field) => new Form.Field({ formDefId: savedDef.id, ...field }, { propertyName: field.propertyName })),
-                  publish
-                );
-            })
-            .then(always(savedDef));
-        })));
+  const fields = await getFormFields(partial.xml);
+  const dataset = await (partial.aux.key.isDefined() ? resolve(Option.none()) : getDataset(partial.xml));
```
[the remark this answers is not in the document]
Comment (ktuite) [reply lib/model/query/forms.js]: done in PR #728

**Review by ktuite — COMMENTED** [review]
comments are from an interactive review with Sadiq earlier today

File: test/unit/data/schema.js [hunk 5]
```diff
@@ -1135,6 +1135,30 @@ describe('form schema', () => {
...
+    it('should say two forms with the same schemas do match', () => Promise.all([
+      fieldsFor(testData.forms.simple),
+      fieldsFor(testData.forms.simple2) // same form structure but different xmlFormId
+    ]).then(([ a, b ]) => {
+      compare(a, b).should.be.true();
+    }));
+
+    it('should say two forms with the different schemas do not match', () => Promise.all([
+      fieldsFor(testData.forms.simple),
+      fieldsFor(testData.forms.withrepeat) // same form structure but different xmlFormId
```
Comment (sadiqkhoja) [inline test/unit/data/schema.js:1148]: comment looks to be incorrect

File: lib/model/migrations/20230109-01-add-form-schema.js [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+*/
+
+const up = async (db) => {
+  // This trigger was responsible for checking for field downcasts that
+  // would break oData streams (e.g. a freeform string becoming an int).
+  // The functionality of it has been recreated in javascript, although
+  // it goes against our design principle of checking constraints in the
+  // database as much as possible. Maybe we'll put it back in the database
+  // someday?
+  await db.raw('DROP TRIGGER check_field_collisions ON form_fields');
```
Comment (sadiqkhoja) [inline lib/model/migrations/20230109-01-add-form-schema.js:32]: How about adding a console log at the start saying that this may take a while?

**Review by sadiqkhoja — APPROVED** [review]
Few minor things, I am happy with the changes

**Pull request merged**